File: student-assignment-02/src/student_assignment_02/src/goal_navigation_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/path.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Vector3.h>
#include <tf2/convert.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <cmath>

using std::placeholders::_1;

class GoalNavigationNode : public rclcpp::Node {
public:
    GoalNavigationNode() : rclcpp::Node("goal_navigation_node") {
        // Subscriber na planiranu putanju
        path_subscription_ = this->create_subscription<nav_msgs::msg::Path>(
            "/planned_path",
            rclcpp::QoS(10),
            std::bind(&GoalNavigationNode::path_callback, this, _1));

        // Publisher za brzinu robota
        cmd_vel_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>(
            "/cmd_vel", 10);

        // Timer za gibanje
        move_timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&GoalNavigationNode::move_callback, this));

        // Parametri
        this->declare_parameter("linear_speed", 0.2);
        this->declare_parameter("angular_speed", 0.5);
        this->declare_parameter("goal_tolerance", 0.1);
        this->declare_parameter("angle_tolerance", 0.1);

        linear_speed_ = this->get_parameter("linear_speed").as_double();
        angular_speed_ = this->get_parameter("angular_speed").as_double();
        goal_tolerance_ = this->get_parameter("goal_tolerance").as_double();
        angle_tolerance_ = this->get_parameter("angle_tolerance").as_double();

        RCLCPP_INFO(this->get_logger(), "Goal Navigation Node inicijaliziran");
        RCLCPP_INFO(this->get_logger(), "  Linear speed: %.2f m/s", linear_speed_);
        RCLCPP_INFO(this->get_logger(), "  Angular speed: %.2f rad/s", angular_speed_);
    }

private:
    rclcpp::Subscription<nav_msgs::msg::Path>::SharedPtr path_subscription_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_publisher_;
    rclcpp::TimerBase::SharedPtr move_timer_;

    nav_msgs::msg::Path current_path_;
    bool path_received_ = false;
    size_t current_waypoint_idx_ = 0;
    bool is_moving_ = false;

    double linear_speed_;
    double angular_speed_;
    double goal_tolerance_;
    double angle_tolerance_;

    void path_callback(const nav_msgs::msg::Path::SharedPtr msg) {
        current_path_ = *msg;
        path_received_ = true;
        current_waypoint_idx_ = 0;
        is_moving_ = true;

        RCLCPP_INFO(this->get_logger(), "Nova putanja primljena s %zu točaka", 
                   current_path_.poses.size());
    }

    void move_callback() {
        if (!path_received_ || !is_moving_ || current_path_.poses.empty()) {
            // Stani
            stop_robot();
            return;
        }

        // Ako smo dostigli zadnji waypoint
        if (current_waypoint_idx_ >= current_path_.poses.size()) {
            RCLCPP_INFO(this->get_logger(), "Cilj dostignut! Gibanje završeno.");
            stop_robot();
            is_moving_ = false;
            path_received_ = false;
            return;
        }

        // Trenutni cilj
        auto target_pose = current_path_.poses[current_waypoint_idx_].pose;
        auto target_x = target_pose.position.x;
        auto target_y = target_pose.position.y;

        // Simulirana pozicija robota (u stvarnoj situaciji biće iz /tf ili /odom)
        // Za sada pretpostavljamo da se robot giba u smjeru putanje
        static double robot_x = 0.0, robot_y = 0.0, robot_theta = 0.0;
        static bool first_call = true;

        if (first_call) {
            if (!current_path_.poses.empty()) {
                robot_x = current_path_.poses[0].pose.position.x;
                robot_y = current_path_.poses[0].pose.position.y;
                first_call = false;
            }
        }

        // Izračun distancije do cilja
        double dx = target_x - robot_x;
        double dy = target_y - robot_y;
        double distance = std::sqrt(dx * dx + dy * dy);

        // Ako je distanca manja od tolerance, idi na sljedeći waypoint
        if (distance < goal_tolerance_) {
            current_waypoint_idx_++;
            RCLCPP_INFO(this->get_logger(), "Waypoint %zu dostignut. Ideći...",
                       current_waypoint_idx_);
            return;
        }

        // Izračun kuta do cilja
        double target_angle = std::atan2(dy, dx);
        double angle_diff = target_angle - robot_theta;

        // Normaliziranje kuta na [-pi, pi]
        while (angle_diff > M_PI) angle_diff -= 2 * M_PI;
        while (angle_diff < -M_PI) angle_diff += 2 * M_PI;

        geometry_msgs::msg::Twist cmd_vel;

        // Ako trebamo rotirati
        if (std::abs(angle_diff) > angle_tolerance_) {
            cmd_vel.linear.x = 0.0;
            cmd_vel.angular.z = (angle_diff > 0) ? angular_speed_ : -angular_speed_;
            
            RCLCPP_DEBUG(this->get_logger(), "Rotacija: angle_diff=%.2f rad", angle_diff);
        } else {
            // Kreni prema cilju
            cmd_vel.linear.x = linear_speed_;
            cmd_vel.angular.z = angular_speed_ * (angle_diff / 0.1);  // PD kontrola

            // Simulacija gibanja robota
            robot_x += linear_speed_ * std::cos(robot_theta) * 0.1;
            robot_y += linear_speed_ * std::sin(robot_theta) * 0.1;
            robot_theta += cmd_vel.angular.z * 0.1;

            if (current_waypoint_idx_ % 5 == 0) {
                RCLCPP_DEBUG(this->get_logger(), 
                           "Gibanje: robot(%.2f, %.2f) -> target(%.2f, %.2f), d=%.2f",
                           robot_x, robot_y, target_x, target_y, distance);
            }
        }

        cmd_vel_publisher_->publish(cmd_vel);
    }

    void stop_robot() {
        geometry_msgs::msg::Twist cmd_vel;
        cmd_vel.linear.x = 0.0;
        cmd_vel.angular.z = 0.0;
        cmd_vel_publisher_->publish(cmd_vel);
    }
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<GoalNavigationNode>());
    rclcpp::shutdown();
    return 0;
}

File: student-assignment-02/src/student_assignment_02/student_assignment_02/goal_navigation_node.py
#!/usr/bin/env python3
"""
Goal Navigation Node
ROS 2 node for robot navigation along planned path
"""

import rclpy
from rclpy.node import Node
import math

from nav_msgs.msg import Path
from geometry_msgs.msg import Twist
from student_assignment_02.utils import normalize_angle


class GoalNavigationNode(Node):
    """ROS 2 Goal Navigation Node"""

    def __init__(self):
        super().__init__('goal_navigation_node')
        
        # Subscriptions
        self.path_sub = self.create_subscription(
            Path,
            '/planned_path',
            self.path_callback,
            rclpy.qos.qos_profile_system_default
        )
        
        # Publishers
        self.cmd_vel_pub = self.create_publisher(
            Twist,
            '/cmd_vel',
            10
        )
        
        # Timer for movement loop
        self.move_timer = self.create_timer(
            0.1,  # 100 ms
            self.move_callback
        )
        
        # Declare parameters
        self.declare_parameter('linear_speed', 0.2)
        self.declare_parameter('angular_speed', 0.5)
        self.declare_parameter('waypoint_tolerance', 0.15)
        self.declare_parameter('angle_tolerance', 0.2)
        
        # Get parameters
        self.linear_speed = self.get_parameter('linear_speed').value
        self.angular_speed = self.get_parameter('angular_speed').value
        self.waypoint_tolerance = self.get_parameter('waypoint_tolerance').value
        self.angle_tolerance = self.get_parameter('angle_tolerance').value
        
        # State
        self.current_path = None
        self.path_received = False
        self.current_waypoint_idx = 0
        self.is_moving = False
        
        # Simulated robot pose
        self.robot_x = 0.0
        self.robot_y = 0.0
        self.robot_theta = 0.0
        
        # Logging
        self.get_logger().info('Goal Navigation Node initialized')
        self.get_logger().info(f'  Linear speed: {self.linear_speed:.2f} m/s')
        self.get_logger().info(f'  Angular speed: {self.angular_speed:.2f} rad/s')
        self.get_logger().info(f'  Waypoint tolerance: {self.waypoint_tolerance:.2f} m')
        self.get_logger().info(f'  Angle tolerance: {self.angle_tolerance:.2f} rad')

    def path_callback(self, msg: Path):
        """Callback for planned path subscription"""
        self.current_path = msg
        self.path_received = True
        self.current_waypoint_idx = 0
        self.is_moving = True
        
        self.get_logger().info(
            f'✅ New path received with {len(msg.poses)} points'
        )
        
        # Initialize robot position at first waypoint
        if msg.poses:
            self.robot_x = msg.poses[0].pose.position.x
            self.robot_y = msg.poses[0].pose.position.y

    def move_callback(self):
        """Timer callback for movement loop"""
        if not self.path_received or not self.is_moving or not self.current_path.poses:
            self.stop_robot()
            return
        
        # Check if all waypoints are reached
        if self.current_waypoint_idx >= len(self.current_path.poses):
            self.get_logger().info('✅ Goal reached! Movement finished.')
            self.stop_robot()
            self.is_moving = False
            self.path_received = False
            return
        
        # Get current target waypoint
        target_pose = self.current_path.poses[self.current_waypoint_idx].pose
        target_x = target_pose.position.x
        target_y = target_pose.position.y
        
        # Calculate distance to target
        dx = target_x - self.robot_x
        dy = target_y - self.robot_y
        distance = math.sqrt(dx * dx + dy * dy)
        
        # Check if waypoint reached
        if distance < self.waypoint_tolerance:
            self.current_waypoint_idx += 1
            if (self.current_waypoint_idx % 10 == 0 or 
                self.current_waypoint_idx == len(self.current_path.poses)):
                self.get_logger().info(
                    f'✅ Waypoint {self.current_waypoint_idx} reached ({distance:.2f} m)'
                )
            return
        
        # Calculate angle to target
        target_angle = math.atan2(dy, dx)
        angle_diff = normalize_angle(target_angle - self.robot_theta)
        
        # Create velocity command
        cmd_vel = Twist()
        cmd_vel.linear.z = 0.0
        cmd_vel.angular.x = 0.0
        cmd_vel.angular.y = 0.0
        
        # Control logic:
        # 1. If angle error is large, rotate in place
        # 2. Otherwise, move forward with small angle correction
        
        if abs(angle_diff) > self.angle_tolerance:
            # Rotate to face target
            cmd_vel.linear.x = 0.0
            cmd_vel.angular.z = (
                self.angular_speed if angle_diff > 0 else -self.angular_speed
            )
        else:
            # Move forward with angle correction
            cmd_vel.linear.x = self.linear_speed
            # PD control for angle
            cmd_vel.angular.z = max(
                -self.angular_speed,
                min(self.angular_speed, angle_diff * 0.5)
            )
        
        # Publish velocity command
        self.cmd_vel_pub.publish(cmd_vel)
        
        # Simulate robot motion
        dt = 0.1  # Timer interval
        self.robot_x += cmd_vel.linear.x * math.cos(self.robot_theta) * dt
        self.robot_y += cmd_vel.linear.x * math.sin(self.robot_theta) * dt
        self.robot_theta = normalize_angle(self.robot_theta + cmd_vel.angular.z * dt)
        
        # Log every 20 iterations
        if self.current_waypoint_idx % 20 == 0:
            self.get_logger().debug(
                f'[WP {self.current_waypoint_idx}] '
                f'Robot: ({self.robot_x:.2f}, {self.robot_y:.2f}, {self.robot_theta:.2f} rad) | '
                f'Target: ({target_x:.2f}, {target_y:.2f}) | '
                f'Distance: {distance:.2f} m | '
                f'Angle diff: {angle_diff:.2f} rad'
            )

    def stop_robot(self):
        """Stop robot by sending zero velocity"""
        cmd_vel = Twist()
        cmd_vel.linear.x = 0.0
        cmd_vel.linear.y = 0.0
        cmd_vel.linear.z = 0.0
        cmd_vel.angular.x = 0.0
        cmd_vel.angular.y = 0.0
        cmd_vel.angular.z = 0.0
        self.cmd_vel_pub.publish(cmd_vel)


def main(args=None):
    """Main function"""
    rclpy.init(args=args)
    node = GoalNavigationNode()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
